File: Java/other/SudokuSolver.java
import java.util.Scanner;

public class SudokuSolver {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter elements of unsolved sudoku. Type 0 for empty cells:");
        int[][] sudokuArr = new int[9][9];
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                sudokuArr[i][j] = scanner.nextInt();
            }
        }
        scanner.close();

        if (solveSudoku(sudokuArr)) {
            System.out.println("Solved sudoku:");
            for (int i = 0; i < 9; i++) {
                for (int j = 0; j < 9; j++) {
                    System.out.print(sudokuArr[i][j]+ ", ");
                }
                System.out.println();
            }
        } else {
            System.out.println("Unsolvable");
        }
    }

    private static boolean solveSudoku(int[][] sudokuArr) {
        int row = -1, col = -1;
        boolean isEmpty = true;

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (sudokuArr[i][j] == 0) {
                    row = i;
                    col = j;

                    // we still have some remaining
                    // missing values in Sudoku
                    isEmpty = false;
                    break;
                }
            }
            if (!isEmpty) {
                break;
            }
        }

        if (isEmpty) {
            return true;
        }

        for (int i = 1; i <= 9; i++) {
            if (isSafe(sudokuArr, row, col, i)) {
                sudokuArr[row][col] = i;
                if (solveSudoku(sudokuArr)) {
                    return true;
                }
                else {
                    sudokuArr[row][col] = 0;
                }
            }
        }

        return false;
    }

    private static boolean isSafe(int[][] sudokuArr, int row, int col, int i) {
        for (int d = 0; d < 9; d++) {
            if (sudokuArr[row][d] == i) {
                return false;
            }
        }

        for (int r = 0; r < 9; r++) {
            if (sudokuArr[r][col] == i) {
                return false;
            }
        }

        int rowStart = row - (row % 3);
        int colStart = col - (col % 3);
        for (int r = rowStart; r < rowStart + 3; r++) {
            for (int d = colStart; d < colStart + 3; d++) {
                if (sudokuArr[r][d] == i) {
                    return false;
                }
            }
        }

        return true;
    }
}

/*
* Sample input/output:
* Enter elements of unsolved sudoku. Type 0 for empty cells:
* 3 0 6 5 0 8 4 0 0
* 5 2 0 0 0 0 0 0 0
* 0 8 7 0 0 0 0 3 1
* 0 0 3 0 1 0 0 8 0
* 9 0 0 8 6 3 0 0 5
* 0 5 0 0 9 0 6 0 0
* 1 3 0 0 0 0 2 5 0
* 0 0 0 0 0 0 0 7 4
* 0 0 5 2 0 6 3 0 0
* Solved sudoku:
* 3, 1, 6, 5, 7, 8, 4, 9, 2,
* 5, 2, 9, 1, 3, 4, 7, 6, 8,
* 4, 8, 7, 6, 2, 9, 5, 3, 1,
* 2, 6, 3, 4, 1, 5, 9, 8, 7,
* 9, 7, 4, 8, 6, 3, 1, 2, 5,
* 8, 5, 1, 7, 9, 2, 6, 4, 3,
* 1, 3, 8, 9, 4, 7, 2, 5, 6,
* 6, 9, 2, 3, 5, 1, 8, 7, 4,
* 7, 4, 5, 2, 8, 6, 3, 1, 9,
*
* Time complexity: O(9^(n*n))
* Space Complexity: O(n*n)
*/

File: C-Plus-Plus/ds/Sudoku_Solver.cpp
// Sudoku Solver

// Write a program to solve a Sudoku puzzle by filling the empty cells.
// A sudoku solution must satisfy all of the following rules:
// Each of the digits 1-9 must occur exactly once in each row.
// Each of the digits 1-9 must occur exactly once in each column.
// Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
// The '.' character indicates empty cells.

// The following code can be used to solve a sudoku which is done using recursion and backtracking.


class Solution {
public:
    bool isSafe(vector<vector<char>> grid, int row,int col, char num)      //we check if block is safe or not
    {
    for (int x = 0; x <= 8; x++){                                           //if no same value is present in the same row
        if (grid[row][x] == num)
            return false;
    }
    for (int x = 0; x <= 8; x++){                                           //if no same value is present in the same column
        if (grid[x][col] == num)
            return false;
    }
    int startRow = row - row % 3,startCol = col - col % 3;
   
    for (int i = 0; i < 3; i++){                                            //if no same value is present in the same 3*3 box
        for (int j = 0; j < 3; j++){
            if (grid[i+startRow][j+startCol] == num)
                return false;
        }
    }
 
    return true;
}
    bool recur(int i,int j,vector<vector<char>>& board){
        if(j==9){                                                           //if j excedes the boundry means we need to move to new row
          i++;  j=0;
        }
        if(i==9)
            return true;
        if(board[i][j]=='.'){                                               //we enter here if block is empty
            for(int k=1;k<=9;k++){                                          //check for all values from [1,9], to see which suites better
                if(isSafe(board,i,j,k+'0')){
                    board[i][j]=char(k+'0');
                    if(recur(i,j+1,board))                                  //if ones true,return same for all previous recursion calls
                        return true;
                    board[i][j]='.';
                }
            }
        }
        else{                                                               //if block is already filled we just move to other bolck
                if(recur(i,j+1,board))
                    return true;
        }
        return false;                                                       //if no value suites the position we return false
    }
    void solveSudoku(vector<vector<char>>& board) {
        recur(0,0,board);
    }
};


// Simple Input and Output
// Input: board = 
//             [["5","3",".",".","7",".",".",".","."],
//             ["6",".",".","1","9","5",".",".","."],
//             [".","9","8",".",".",".",".","6","."],
//             ["8",".",".",".","6",".",".",".","3"],
//             ["4",".",".","8",".","3",".",".","1"],
//             ["7",".",".",".","2",".",".",".","6"],
//             [".","6",".",".",".",".","2","8","."],
//             [".",".",".","4","1","9",".",".","5"],
//             [".",".",".",".","8",".",".","7","9"]]
<p align="center">
  <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/ff/Sudoku-by-L2G-20050714.svg/250px-Sudoku-by-L2G-20050714.svg.png"  />
</p>

// Output: 
//             [["5","3","4","6","7","8","9","1","2"],
//             ["6","7","2","1","9","5","3","4","8"],
//             ["1","9","8","3","4","2","5","6","7"],
//             ["8","5","9","7","6","1","4","2","3"],
//             ["4","2","6","8","5","3","7","9","1"],
//             ["7","1","3","9","2","4","8","5","6"],
//             ["9","6","1","5","3","7","2","8","4"],
//             ["2","8","7","4","1","9","6","3","5"],
//             ["3","4","5","2","8","6","1","7","9"]]

<p align="center">
  <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/31/Sudoku-by-L2G-20050714_solution.svg/250px-Sudoku-by-L2G-20050714_solution.svg.png"  />
</p>

// Explanation: The input board is shown above and the only valid solution is shown


// Here we have used the same logic as that of N-queens problem
// Time taken will be O(9^(n*n))
// Space complexity O(n*n)